app/src/main/java/org/shadowice/flocke/andotp/Activities/AuthenticateActivity.java

```diff
@@ -22,190 +22,245 @@
-
-                if (key == null)
-                    Toast.makeText(this, R.string.settings_toast_auth_upgrade_failed, Toast.LENGTH_LONG).show();
-
-                if (authMethod == AuthMethod.PASSWORD)
-                    settings.removeAuthPasswordHash();
-                else if (authMethod == AuthMethod.PIN)
-                    settings.removeAuthPINHash();
+    private void startAuthTask(String plainPassword) {
+        TaskFragment taskFragment = findTaskFragment();
+        // Don't start another task if this was already started.
+        if (taskFragment == null) {
+            // Add a retained task fragment so that we will be able to know if a task was running on a configuration change.
+            AuthenticationTask task = new AuthenticationTask(this, isAuthUpgrade, existingAuthCredentials, plainPassword);
+            task.setCallback(this::handleResult);
+            taskFragment = new TaskFragment();
+            taskFragment.startTask(task);
+            getFragmentManager()
+                    .beginTransaction()
+                    .add(taskFragment, TAG_TASK_FRAGMENT)
+                    .commit();
+            setupUiForRunningTask();
+        }
+    }
 
-                finishWithResult(true, key);
-            } else {
-                finishWithResult(false, null);
-            }
+    private void handleResult(Result result) {
+        if (result.authUpgradeFailed) {
+            Toast.makeText(this, R.string.settings_toast_auth_upgrade_failed, Toast.LENGTH_LONG).show();
         }
+        finishWithResult(result.encryptionKey != null, result.encryptionKey);
     }
 
-    // End with a result
-    public void finishWithResult(boolean success, byte[] key) {
+    private void finishWithResult(boolean success, byte[] encryptionKey) {
         Intent data = new Intent();
-
-        if (newEncryption != null && ! newEncryption.isEmpty())
+        if (newEncryption != null && !newEncryption.isEmpty())
             data.putExtra(Constants.EXTRA_AUTH_NEW_ENCRYPTION, newEncryption);
-
-        if (key != null)
-            data.putExtra(Constants.EXTRA_AUTH_PASSWORD_KEY, key);
-
+        if (encryptionKey != null)
+            data.putExtra(Constants.EXTRA_AUTH_PASSWORD_KEY, encryptionKey);
         if (success)
             setResult(RESULT_OK, data);
-
         finish();
     }
 
-    // Go back to the main activity
+    @Override
+    protected void onPause() {
+        super.onPause();
+        // We don't want the task to callback to a dead activity and cause a memory leak, so null it here.
+        TaskFragment taskFragment = findTaskFragment();
+        if (taskFragment != null) {
+            taskFragment.task.setCallback(null);
+        }
+    }
+
     @Override
     public void onBackPressed() {
         finishWithResult(false, null);
         super.onBackPressed();
     }
+
+    /** Retained instance fragment to hold a running @link{{@link AuthenticationTask}} between configuration changes.*/
```

Comment:
Just noticed this weird double javadoc link, I'll fix that real quick on my lunch break